bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+package org.mozartoz.bootcompiler
+package transform
+
+import ast._
+import oz._
+import symtab._
+import scala.collection._
+import com.sun.javafx.fxml.expression.BinaryExpression
+
+sealed class VarState(val onStack: Boolean, val locked: Boolean) {
+}
+object Declared      extends VarState(false, false)
+object Captured      extends VarState(false, false)
+object BoundFirst    extends VarState(true, true)
+object ToInstantiate extends VarState(false, true)
+
+case class VarAndState(v: Variable, state: VarState)
+
+/**
+ * Looks for declarations and the usages of those declarations.
+ * The Transform methods detect local statements and add declared variables to the environment
+ * The Walk methods are collecting their usages and tagging them
+ * 
+ * If a variable is bound after its declaration, we can shortcut it
+ * 
+ * When a proc is declared, it is difficult to track where it will end up. So if a call occurs,
+ * by precaution, all captured variable that were not bound should be instantiated
```

Comment:
I think we could consider a call "reading" its arguments and the additional rule that if a variable refers to a proc literal (needs some extra tracking) then it also "reads" all captured variables.
But indeed, that's for another time.
```oz
local A B C in
  A=proc {$} {Show C} end
  B=A
  {B} % must instantiate C
end
```
Alternatively, maybe we could just mark all captured variables by `proc` literal as needing to be instantiated if they are not already bound. That would be trivial but unclear how it would perform compared to the current approach. Calls would not need to be special anymore.